appcues/src/main/java/com/appcues/trait/appcues/TooltipTrait.kt

```diff
@@ -330,38 +353,61 @@ internal class TooltipTrait(
         )
     }
 
-    private fun Modifier.tooltipSize(style: ComponentStyle?, maxWidth: Dp, maxHeight: Dp): Modifier = then(
-        if (style?.width == null) Modifier.requiredWidthIn(
-            min = 48.dp,
-            max = min(400.dp, maxWidth)
-        ) else Modifier.width(width = style.width.dp)
-    ).then(
-        if (style?.height == null) Modifier.requiredHeightIn(
-            min = 48.dp,
-            max = max(48.dp, maxHeight)
-        ) else Modifier.height(height = style.height.dp)
-    )
+    private fun Modifier.tooltipSize(style: ComponentStyle?, windowInfo: AppcuesWindowInfo, tooltipSettings: TooltipSettings): Modifier =
+        composed {
+            // keep containerMaxSize here as a rememberable so it force recomposition when value change
+            val containerMaxSize = DpSize(
+                width = windowInfo.widthDp - (SCREEN_HORIZONTAL_PADDING * 2),
+                height = windowInfo.heightDp - (SCREEN_VERTICAL_PADDING * 2)
+            )
+
+            val width = if (style?.width != null) style.width.dp else 400.dp
+            val pointerLengthDp = with(LocalDensity.current) { tooltipSettings.pointerLengthPx.toDp() }
+
+            val modifier = when (tooltipSettings.tooltipPointerPosition) {
+                Bottom, Top -> Modifier.width(min(width, containerMaxSize.width))
+                Left, Right -> Modifier.width(min(width + pointerLengthDp, containerMaxSize.width))
+                None -> Modifier.width(min(width, containerMaxSize.width))
+            }.requiredHeightIn(48.dp, containerMaxSize.height)
+
+            return@composed modifier
+        }
 
     @Composable
-    private fun rememberTooltipSettings(
+    private fun rememberTooltipPointerPosition(
         windowInfo: AppcuesWindowInfo,
         targetRectangleInfo: TargetRectangleInfo?,
-        containerDimens: TooltipContainerDimens?,
+        contentDimens: TooltipContainerDimens?,
         targetRect: Rect?,
         distance: Dp,
-        tooltipMaxHeight: MutableState<Dp>,
-    ): TooltipSettings {
-        val density = LocalDensity.current
-        return remember(targetRectangleInfo, containerDimens) {
-            val position = targetRectangleInfo.getTooltipPointerPosition(
+        pointerLength: Dp,
+        contentSized: Boolean,
+    ): TooltipPointerPosition {
+        // pointer position is only calculated once for a given target rectangle, after the container
+        // is sized. The tooltip may end up getting a new container layout if it shifts the pointer
+        // to another edge (not preferred edge) based on this calculation, but we stick with that
+        // calculated position from that point forward and don't allow it to flip back
+        return remember(targetRectangleInfo, contentSized) {
```

Comment:
I'm not sure if this remember is 1000% valid, but maybe good enough. Since `contentSized` gets set to true once the first step runs its layout and sets a size - but what if the next step uses the exact same `targetRectangleInfo`? Maybe that is not possible - but what I'm trying to get at: every step should run this pointer position code once, picking the desired position for that step, and then it only resets when a new step comes through (or some other significant relayout happens like an orientation change?)

Reply:
its not intuitive but the targetRectangle cant be the same for two steps I dont think, but in any case, if both steps would share the same, then the contentSized would trigger again when the content size changes

Reply:
I don't think `contentSized` would trigger again though would it? since it was already sized and moved from `false` -> `true` on the first ever layout. it would never change again (since never `null` again)

Reply:
since contentDimens is a key to a remember in that block, when steps change, the size will most likely change as well, and it will cause the entire block to recompose. thats my assumption